## BERT embeddings pour l'analyse textuelle 
#### Utilisation de la librairie Hugging Face
- 1) exploration du contenu d'un modèle BERT (français)
- 2) réglage fin d'un modèle BERT français sur notre corpus
- 3) clustering de documents et topics dérivés (à la Bertopic)
- 4) topics comme clustering de (sens de) tokens (à faire) 

##### Installation des librairies HuggingFace non installées par défaut dans Colab

In [ ]:
!pip install datasets

     |████████████████████████████████| 194kB 6.2MB/s 
     |████████████████████████████████| 245kB 7.6MB/s 
     |████████████████████████████████| 112kB 9.6MB/s 


In [ ]:
!pip install transformers

     |████████████████████████████████| 2.1MB 5.9MB/s 
     |████████████████████████████████| 890kB 24.7MB/s 
     |████████████████████████████████| 3.2MB 41.2MB/s 
  Created wheel for sacremoses: filename=sacremoses-0.0.43-cp37-none-any.whl size=893262 sha256=f80da26e5c4d13769a2a5a8dbe4a02958c1830714bde9b9fdc109eb13fad3bb5
  Stored in directory: /root/.cache/pip/wheels/29/3c/fd/7ce5c3f0666dab31a50123635e6fb5e19ceb42ce38d4e58f45
Successfully built sacremoses


In [ ]:
# pour sauvegarder un graphique spacy sous forme de fichier
!pip install plotly==4.7.1
!wget https://github.com/plotly/orca/releases/download/v1.2.1/orca-1.2.1-x86_64.AppImage -O /usr/local/bin/orca
!chmod +x /usr/local/bin/orca
!apt-get install xvfb libgtk2.0-0 libgconf-2-4

     |████████████████████████████████| 11.5MB 339kB/s 
  Found existing installation: plotly 4.4.1
    Uninstalling plotly-4.4.1:
      Successfully uninstalled plotly-4.4.1
--2021-03-17 11:54:29--  https://github.com/plotly/orca/releases/download/v1.2.1/orca-1.2.1-x86_64.AppImage
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/99037241/9dc3a580-286a-11e9-8a21-4312b7c8a512?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210317%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210317T115429Z&X-Amz-Expires=300&X-Amz-Signature=051601889547ec6cded9bd51af84f43c75c7122cd9fdd238480d7b26ba3523ac&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=99037241&response-content-disposition=attachment%3B%20filename%3Dorca-1.2.1-x86_64.AppImage&response-content-type=application%2Foctet-stream [following

In [ ]:
!pip install hdbscan

     |████████████████████████████████| 6.4MB 4.5MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
  Created wheel for hdbscan: filename=hdbscan-0.8.27-cp37-cp37m-linux_x86_64.whl size=2311674 sha256=6457b9a8e7c54acad9f15dc608cbd9fd74bc4427cda7bd874036e096fc09f4b4
  Stored in directory: /root/.cache/pip/wheels/42/63/fb/314ad6c3b270887a3ecb588b8e5aac50b0fad38ff89bb6dff2
Successfully built hdbscan


## ICI IL FAUT REDEMARRER L'ENVIRONNEMENT D'EXECUTION (pour employer les transformers (et hdbscan))

In [ ]:
# 1.a
# librairies générales (on ne se servira pas forcément de toutes)
import os
import csv
import re
import math
import pickle
import datetime as dt
from ast import literal_eval
import json 
from random import sample
import collections

In [ ]:
# 1.b
# librairies générales data science
import numpy as np
import scipy.stats as ss
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

In [ ]:
# 1.c
# pour affichage dtypes pandas quand plus de 60 colonnes
pd.set_option('display.max_rows', 120)

In [ ]:
# 1.d
# la librairie graphique la plus simple d'utilisation
import plotly.express as px
import plotly.graph_objects as go

In [ ]:
# 1.f réduction de dimension via UMAP
import umap

In [ ]:
# 1.g clustering de densité (soft) via HDBSCAN
import hdbscan

In [ ]:
# 1.h imports spécifiques à google colab
from google.colab import files, drive

In [ ]:
# 1.i passage obligé pour récupérer un fichier sur le drive
# il va falloir choisir le drive (account Google) puis renseigner un mot de passe généré
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 1.j utilisation des transformers HuggingFace dans le cadre de pyTorch
import torch

In [ ]:
# 1.k formatage des données pour entrée dans tokenisateurs HuggingFace
from datasets import Dataset

In [ ]:
# 1.l utilisation des transformers
from transformers import AutoTokenizer
from transformers import AutoModelForMaskedLM

In [ ]:
# 1.m utilisation pour le réglage fin du transformer
from transformers import Trainer, TrainingArguments
from transformers import DataCollatorForLanguageModeling

In [ ]:
# 1.n utilisation pour la mise en oeuvre du modèle réglé finement (un modèle FlauBERT)
from transformers import FlaubertModel

## Structure interne des tokeniseurs et transformers

### Choix du modèle BERT
##### Pour le français, il y a actuellement deux familles de modèles CamemBERT et FlauBERT dont les niveaux de performance sont très voisins, et supérieurs à ceux des modèle multi-lingues (mBERT) incluant le français
##### CamemBERT a demandé plus de ressources à l'entraînement (modèle RoBERTa)
##### Un modèle de base est ici suffisant, et uncased vu nos textes de départ, déjà en minuscules pour l'essentiel
##### *flaubert-base-uncased*

In [ ]:
# 2.a
nom_modele = 'flaubert/flaubert_base_uncased' 

#### Tokenisation

In [ ]:
# 2.b
# tokeniseur : on essaye le chargement de la version rapide du tokeniseur (si le temps est déjà bon, essayer la version plus lente)
tokeniseur = AutoTokenizer.from_pretrained(nom_modele, use_fast=True)

In [ ]:
# 2.c
# essai sur un texte
#texte = "Le malade a pris de l'hydroxychloroquine, et il aurait effectivement guéri du COVID"
texte = "Macron a rencontré le professeur Raoult, la discussion a duré 2 heures."

###### Le tokénisateur est celui de XLM
- caractères spéciaux : s (Beginning Of Sentence), /s (Separator)

###### tokenize renvoie la liste des tokens (sous forme de texte)
- il n'y a pas de séparateurs de début et de fin de bloc
- les mots pleins (formes fléchies incluses) sont sous la forme chaîne du mot suivi de /w (fin de mot)
- pour les mots hors vocabulaire, il y a N sous-tokens, le dernier étant suivi de /w

In [ ]:
# 2.d
texte_tokenise = tokeniseur.tokenize(texte)
print(texte_tokenise)
print(len(texte_tokenise))

['macron</w>', 'a</w>', 'rencontré</w>', 'le</w>', 'professeur</w>', 'raou', 'lt</w>', ',</w>', 'la</w>', 'discussion</w>', 'a</w>', 'duré</w>', '2</w>', 'heures</w>', '.</w>']
15


###### encode renvoie la liste des ids de token
- y compris les séparateurs de début et de fin de bloc (0 et 1)

In [ ]:
# 2.e
texte_encode = tokeniseur.encode(texte)
print(texte_encode)
print(len(texte_encode))

[0, 3898, 34, 3467, 19, 1841, 36915, 7177, 14, 17, 3120, 34, 10029, 122, 363, 16, 1]
17


###### convert_ids_to_tokens renvoie la liste des textes de tokens 
- y compris les début (s) et fin (/s) de bloc

In [ ]:
# 2.f
l_tokens = tokeniseur.convert_ids_to_tokens(texte_encode)
print(l_tokens)

['<s>', 'macron</w>', 'a</w>', 'rencontré</w>', 'le</w>', 'professeur</w>', 'raou', 'lt</w>', ',</w>', 'la</w>', 'discussion</w>', 'a</w>', 'duré</w>', '2</w>', 'heures</w>', '.</w>', '</s>']


###### convert_tokens_to_string revient au texte initial
- avec en plus les débuts (s) et fins (/s) de bloc

In [ ]:
# 2.g
texte_reconstitue = tokeniseur.convert_tokens_to_string(l_tokens)
print(texte_reconstitue)

<s>macron a rencontré le professeur raoult , la discussion a duré 2 heures . </s>


###### encode_plus renvoie un dictionnaire avec 3 éléments :
- input_ids : la liste des ids de tokens
- token_type_ids : la liste des types de tokens (que des 0 dans le cas standard, les 1 peuvent servir à distinguer une phrase de rôle différent, par exemple question vs réponse)
- attention_mask : la liste des masques d'attention (que des 1 si on ne traite qu'un bloc à la fois, on regarde tous les tokens du bloc ; si dans un batch il y a plusieurs blocs, la liste est mise à la longueur la plus grande, et pour les plus courts, on complète avec des 0)
- début et fin de bloc sont représentés dans les 3 éléments

In [ ]:
# 2.h
texte_encode_plus = tokeniseur.encode_plus(texte)
print(len(texte_encode_plus))
for k, v in texte_encode_plus.items():
  print(k, ":", v)

3
input_ids : [0, 3898, 34, 3467, 19, 1841, 36915, 7177, 14, 17, 3120, 34, 10029, 122, 363, 16, 1]
token_type_ids : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
attention_mask : [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


##### utiliser directement le tokéniesur est équivalent à encode_plus

In [ ]:
# 2.i
tokenise = tokeniseur(texte)
print(len(tokenise))
for k, v in tokenise.items():
  print(k, ":", v)

3
input_ids : [0, 3898, 34, 3467, 19, 1841, 36915, 7177, 14, 17, 3120, 34, 10029, 122, 363, 16, 1]
token_type_ids : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
attention_mask : [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


###### On peut aussi garantir un format tenseur (ici pyTorch) en sortie du tokénisateur

In [ ]:
# 2.j
texte_tokenise_pt = tokeniseur(texte, return_tensors="pt")
for k, v in texte_tokenise_pt.items():
  print(k, ":", v, "--- shape :", v.shape)

input_ids : tensor([[    0,  3898,    34,  3467,    19,  1841, 36915,  7177,    14,    17,
          3120,    34, 10029,   122,   363,    16,     1]]) --- shape : torch.Size([1, 17])
token_type_ids : tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]) --- shape : torch.Size([1, 17])
attention_mask : tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]) --- shape : torch.Size([1, 17])


###### >>> batch = tokenizer(batch_sentences, padding=True, truncation=True, return_tensors="pt")
###### >>> print(batch)
{'input_ids': tensor([[ 101, 8667,  146,  112,  182,  170, 1423, 5650,  102],
                      [ 101, 1262, 1330, 5650,  102,    0,    0,    0,    0],
                      [ 101, 1262, 1103, 1304, 1304, 1314, 1141,  102,    0]]),
 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0],
                           [0, 0, 0, 0, 0, 0, 0, 0, 0],
                           [0, 0, 0, 0, 0, 0, 0, 0, 0]]),
 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1],
                           [1, 1, 1, 1, 1, 0, 0, 0, 0],
                           [1, 1, 1, 1, 1, 1, 1, 1, 0]])}

###### Vocabulaire du tokénisateur

In [ ]:
# 2.k
# longueur du vocabulaire
l_vocab_tokeniseur = list(tokeniseur.get_vocab())
print(len(l_vocab_tokeniseur))

67542


In [ ]:
# 2.l
# quelques éléments du vocabulaire
print(l_vocab_tokeniseur[4000:4030])

['fleuries</w>', 'œur</w>', 'ṱ', 'sies</w>', 'xiie</w>', 'mit</w>', 'stimule</w>', 'meur', 'intéressants</w>', 'clai', 'dépassent</w>', 'levi', 'vants</w>', 'sommaire</w>', 'oca</w>', 'raill', 'kaf', 'immatricul', 'ҳ</w>', 'chall', 'alexander</w>', 'antérieurs</w>', 'revanche</w>', 'primer</w>', 'oire</w>', '莎', 'oui', 'rcs</w>', 'motivée</w>', '執']


#### Modèle BERT (FlauBERT)

In [ ]:
# 3.a
# charger le modèle adapté à la prédiction de tokens en contexte
# on aura besoin d'examiner les couches en sortie des derniers transformers, pas seulement le dernier
modele = AutoModelForMaskedLM.from_pretrained(nom_modele, output_hidden_states = True)

Some weights of FlaubertWithLMHeadModel were not initialized from the model checkpoint at flaubert/flaubert_base_uncased and are newly initialized: ['transformer.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
# 3.b
# se mettre en mode exclusivement propagation avancée
# en fait, lorsqu'on vient de charger le modèle via from_pretrained, le modèle est déjà en mode "évaluation" (par opposition à "entraînement")
modele.eval()
# renvoie la liste des composantes (couches) du modèle

FlaubertWithLMHeadModel(
  (transformer): FlaubertModel(
    (position_embeddings): Embedding(512, 768)
    (embeddings): Embedding(67542, 768, padding_idx=2)
    (layer_norm_emb): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (attentions): ModuleList(
      (0): MultiHeadAttention(
        (q_lin): Linear(in_features=768, out_features=768, bias=True)
        (k_lin): Linear(in_features=768, out_features=768, bias=True)
        (v_lin): Linear(in_features=768, out_features=768, bias=True)
        (out_lin): Linear(in_features=768, out_features=768, bias=True)
      )
      (1): MultiHeadAttention(
        (q_lin): Linear(in_features=768, out_features=768, bias=True)
        (k_lin): Linear(in_features=768, out_features=768, bias=True)
        (v_lin): Linear(in_features=768, out_features=768, bias=True)
        (out_lin): Linear(in_features=768, out_features=768, bias=True)
      )
      (2): MultiHeadAttention(
        (q_lin): Linear(in_features=768, out_features=768, bia

In [ ]:
# 3.c
# mettre en entrée le texte tokenizé, et récupérer la sortie
sorties = modele(**texte_tokenise_pt)

###### logits (torch.FloatTensor of shape (batch_size, sequence_length, config.vocab_size)) – Prediction scores of the language modeling head (scores for each vocabulary token before SoftMax)
###### hidden_states Tuple of torch.FloatTensor (one for the output of the embeddings + one for the output of each layer) of shape (batch_size, sequence_length, hidden_size)

In [ ]:
# 3.d
sorties_logits = sorties['logits']
print(type(sorties_logits))
print(len(sorties_logits))
print(sorties_logits.shape)

<class 'torch.Tensor'>
1
torch.Size([1, 17, 67542])


In [ ]:
# 3.e
sorties_transformers = sorties['hidden_states']
print(type(sorties_transformers))
print(len(sorties_transformers))
for l in range(len(sorties_transformers)):
  print(type(sorties_transformers[l]))
  print(l, sorties_transformers[l].shape)

<class 'tuple'>
13
<class 'torch.Tensor'>
0 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
1 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
2 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
3 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
4 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
5 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
6 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
7 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
8 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
9 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
10 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
11 torch.Size([1, 17, 768])
<class 'torch.Tensor'>
12 torch.Size([1, 17, 768])


In [ ]:
# 3.f
# récupérer le 1er élément du 1er tenseur, et le tout dernier
print(sorties_transformers[0][0, 0, 0])
print(sorties_transformers[12][0, 16, 767])

tensor(-0.2951, grad_fn=<SelectBackward>)
tensor(0.6149, grad_fn=<SelectBackward>)


#### Sélectionner / reformater sorties des transformers pour une exploitation facilitée des couches
###### Notamment pour extraire un embedding représentatif de la transformation des tokens

In [ ]:
# 3.g
# concaténer toutes les couches
embeddings_tokens = torch.stack(sorties_transformers, dim=0)
print(embeddings_tokens.shape)

torch.Size([13, 1, 17, 768])


In [ ]:
# 3.h
# éliminer la dimension batch (n° 1 maintenant)
embeddings_tokens = torch.squeeze(embeddings_tokens, dim=1)
print(embeddings_tokens.shape)

torch.Size([13, 17, 768])


In [ ]:
# 3.i
# permuter dimensions des couches et des tokens
embeddings_tokens = embeddings_tokens.permute(1, 0, 2)
print(embeddings_tokens.shape)

torch.Size([17, 13, 768])


In [ ]:
# 3.j
# on peut transformer le tout en numpy array
mx_embeddings_tokens = embeddings_tokens.cpu().detach().numpy()
print(mx_embeddings_tokens.shape)

(17, 13, 768)


###### Obtention des vecteurs de tokens, puis de phrases

In [ ]:
# 3.k
# une stratégie classique est de moyenner les N dernières couches
NB_DERNIERES_COUCHES = 4
ix_dernieres_couches = 13 - NB_DERNIERES_COUCHES

In [ ]:
# 3.l
# sélection des dernières couches
mx_sel_N_dernieres_couches = mx_embeddings_tokens[:, ix_dernieres_couches:, :]
print(mx_sel_N_dernieres_couches.shape)

(17, 4, 768)


In [ ]:
# 3.m
# pour chaque token (1ère dimension) il faut sommer sur les N couches (2ème dimension), en gardant les embeddings (dernière dimension)
# les tokens sont alors en ligne, et les embeddings toujours en colonne
mx_somme_N_dernieres_couches = np.sum(mx_sel_N_dernieres_couches, axis=1)
print(mx_somme_N_dernieres_couches.shape)

(17, 768)


In [ ]:
# 3.n
# pour un embedding de sentence, on réduit encore en sommant sur l'axe 0 (celui des tokens, qu'on élimine)
mx_somme_sur_bloc = np.sum(mx_somme_N_dernieres_couches, axis=0)
print(mx_somme_sur_bloc.shape)

(768,)


#### Traitement en batch de plusieurs textes

In [ ]:
# 3.o
# ici deux textes, de longueurs différentes
l_textes = ["Macron a rencontré le professeur Raoult, la discussion a duré 2 heures.", "Vraiment ?"]

###### Comme on a des textes de différentes longueurs, il faut demander de compléter la suite des textes les moins courts (padding=True). Il est aussi plus prudent de demander la troncation des textes les plus longs pour qu'ils puissent entrer dans le transformer

In [ ]:
# 3.p
# tokénisation de la liste de textes
l_textes_tokenises_pt = tokeniseur(l_textes, padding=True, truncation=True, return_tensors="pt")
for k, v in l_textes_tokenises_pt.items():
  print(k, ":", v, "--- shape :", v.shape)

input_ids : tensor([[    0,  3898,    34,  3467,    19,  1841, 36915,  7177,    14,    17,
          3120,    34, 10029,   122,   363,    16,     1],
        [    0,   367,    88,     1,     2,     2,     2,     2,     2,     2,
             2,     2,     2,     2,     2,     2,     2]]) --- shape : torch.Size([2, 17])
token_type_ids : tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]) --- shape : torch.Size([2, 17])
attention_mask : tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]) --- shape : torch.Size([2, 17])


In [ ]:
# 3.q
# on peut pour chaque texte récupérer le nb de tokens avant padding
l_len_textes = []
for no_texte in range(len(l_textes)):
  masque_attention = l_textes_tokenises_pt['attention_mask'][no_texte]
  nb_tokens = masque_attention.shape[0]
  len_texte = 0
  for m in range(nb_tokens):
    if masque_attention[m] == 1: len_texte += 1
    else: break
  l_len_textes.append(len_texte)
print(l_len_textes)

[17, 4]


In [ ]:
# 3.r
# mettre en entrée les textes tokenizés, et récupérer la sortie
sorties = modele(**l_textes_tokenises_pt)

##### 13 couches cachées de format [nb de textes du batch, nb de tokens pour chaque texte, nb de dimensions d'embedding]

In [ ]:
# 3.s
# examen des couches cachées
sorties_transformers = sorties['hidden_states']
print(type(sorties_transformers))
print(len(sorties_transformers))
for l in range(len(sorties_transformers)):
  print(type(sorties_transformers[l]))
  print(l, sorties_transformers[l].shape)

<class 'tuple'>
13
<class 'torch.Tensor'>
0 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
1 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
2 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
3 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
4 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
5 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
6 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
7 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
8 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
9 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
10 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
11 torch.Size([2, 17, 768])
<class 'torch.Tensor'>
12 torch.Size([2, 17, 768])


In [ ]:
# 3.t
# concaténer toutes les couches
embeddings_tokens = torch.stack(sorties_transformers, dim=0)
print(embeddings_tokens.shape)

torch.Size([13, 2, 17, 768])


In [ ]:
# 3.u
# permuter les dimensions de façon à avoir l'ordre batch, tokens, couches et embeddings
embeddings_tokens = embeddings_tokens.permute(1, 2, 0, 3)
print(embeddings_tokens.shape)

torch.Size([2, 17, 13, 768])


In [ ]:
# 3.v
# on peut transformer le tout en numpy array
mx_embeddings_tokens = embeddings_tokens.cpu().detach().numpy()
print(mx_embeddings_tokens.shape)

(2, 17, 13, 768)


In [ ]:
# 3.w
# une stratégie classique est de moyenner les N dernières couches
NB_DERNIERES_COUCHES = 4
ix_dernieres_couches = 13 - NB_DERNIERES_COUCHES

In [ ]:
# 3.x
# boucler texte par texte (sur la dimension batch)
nb_textes = mx_embeddings_tokens.shape[0]
for no_texte in range(nb_textes):
  print(no_texte)

  # extraire l'embedding du texte
  mx_texte_embeddings_tokens = mx_embeddings_tokens[no_texte]

  # sélection des dernières couches
  mx_sel_N_dernieres_couches = mx_texte_embeddings_tokens[:, ix_dernieres_couches:, :]
  print(mx_sel_N_dernieres_couches.shape)

  # pour chaque token (1ère dimension) il faut sommer sur les N couches (2ème dimension), en gardant les embeddings (dernière dimension)
  # les tokens sont alors en ligne, et les embeddings toujours en colonne
  mx_somme_N_dernieres_couches = np.sum(mx_sel_N_dernieres_couches, axis=1)
  print(mx_somme_N_dernieres_couches.shape)

  # vérification de l'effet du padding (le second texte fait 4 de long)
  for i in [0, 1, 2, 3, 4, 15, 16]:
    print(i, mx_somme_N_dernieres_couches[i][0], mx_somme_N_dernieres_couches[i][767])

  # pour un embedding de sentence, on réduit encore en sommant sur l'axe 0 (celui des tokens, qu'on élimine)
  mx_somme_sur_bloc = np.sum(mx_somme_N_dernieres_couches, axis=0)
  print(mx_somme_sur_bloc.shape)



0
(17, 4, 768)
(17, 768)
0.22845879 2.6680057
-0.73210144 2.6488466
-1.3033249 1.7048044
-0.7059021 -1.8785855
0.81676507 0.32797867
-0.14851315 0.57719284
(768,)
1
(17, 4, 768)
(17, 768)
2.2612886 3.9173677
1.8151124 4.400249
2.1075423 4.393114
0.0 0.0
0.0 0.0
0.0 0.0
(768,)


## Lecture du fichier de 1300 vidéos prétraitées
##### Prétraitées : on a effectué au préalable l'analyse nlp de Spacy sur les sous_titres et renseigné :
###### 'textes_sous_titres_avec_bigrammes', texte ne contenant que les lemmes pleins, éventuellement regroupés en bigrammes pertinents
###### en plus de ce qui avait été précédemment utilisé
###### 'langue' langue de la vidéo tel que détecté par spacy (on garde 'fr')
###### 'tkn_sous_titres' listes de token avec un tuple par token
###### 'ent_sous_titres' liste d'entités nommées avec un tuple par entité nommée

In [ ]:
# 4.a sous-répertoire éventuellement à adapter
chemin_nom_dataframe = "/content/drive/My Drive/Datasets NLP/COVID 19/videos_covid_nlp_2.csv"

In [ ]:
# 4.b
df_videos = pd.read_csv(chemin_nom_dataframe, sep=',', encoding='utf-8')
nb_videos = len(df_videos)
print(nb_videos)

1342


In [ ]:
# 4.c
df_videos.dtypes

video_id                              object
channel_id                            object
titre                                 object
description                           object
date_publication                      object
durée                                 object
catégorie_YT                         float64
tags                                  object
nb_vues                              float64
nom_chaîne                            object
sous_titres                           object
nb_mots                              float64
mode_création                         object
nb_vues_chaîne                         int64
catégorie_SP                          object
langue_sous_titres                    object
tkn_sous_titres                       object
ent_sous_titres                       object
l_textes_l_tokens                     object
textes_sous_titres_avec_bigrammes     object
dtype: object

In [ ]:
# 4.d
nb_docs = nb_videos

###### Auparavant, normaliser la colonnes sous-titres (enlever les [MUSIQUE] et autres informations sonores)
###### Il faut aussi passer les éventuelles majuscules en minuscules, car les sous-titres sont à la base en minuscule, et on va utiliser un modèle BERT en minuscules

In [ ]:
# 4.e
# mais avant normaliser les sous-titres
# éliminer [Musique] et chaînes analogues...
FOND_SONORE = re.compile('\[\w+\]')
def normalise_sous_titre(text):
    text_normalise = re.sub(FOND_SONORE, "", text)
    return text_normalise


In [ ]:
# 4.f
df_videos['texte_base'] = df_videos['sous_titres'].apply(lambda x: normalise_sous_titre(x).lower())

## Réglage fin du modèle Flaubert

#### Pour cette partie, il faut passer le notebook en mode GPU
###### Pour ce, aller dans Modifier | Paramètres du notebook, choisir GPU comme Accélérateur matériel 

##### En préalable, créer un dataset avec un jeu d'entraînement et un jeu de test
###### On utilise la colonne sous-titres de texte d'origine, car les embeddings BERT (et les autres d'ailleurs) prennent en compte tous les mots (stop-words compris) et sous leurs formes fléchies (les flexions étant prises en compte au niveau des subtokens)

##### Pour le réglage-fin, on est uniquement intéressé par la colonne de texte qui va être utilisé pour
##### On crée un dataset HuggingFace à partir du dataframe pandas
##### Et dessus, on fait une séparation entre textes d'entraînement et textes de validation, par une méthode inspirée de scikit-learn 

In [ ]:
# 5.a
df_texte = df_videos[['texte_base']]

In [ ]:
# 5.b
ds_texte = Dataset.from_pandas(df_texte)

In [ ]:
# 5.c
# l'objet créé est un dictionnaire de datasets, avec un dataset 'train' et un dataset 'test'
# on a pris 80 % des données pour l'entraînement, 20 % pour le test
# lorsque le nombre de textes est grand, on peut en réserver plus pour l'entraînement
ds_texte = ds_texte.train_test_split(test_size=0.2)
ds_train = ds_texte['train']
ds_test = ds_texte['test']

##### Vérification, avec fonction directement recopiée d'un notebook Hugging Face

In [ ]:
# 5.d
from datasets import ClassLabel
import random
from IPython.display import display, HTML

def show_random_elements(dataset, num_examples=10):
    assert num_examples <= len(dataset), "Can't pick more elements than there are in the dataset."
    picks = []
    for _ in range(num_examples):
        pick = random.randint(0, len(dataset)-1)
        while pick in picks:
            pick = random.randint(0, len(dataset)-1)
        picks.append(pick)
    
    df = pd.DataFrame(dataset[picks])
    for column, typ in dataset.features.items():
        if isinstance(typ, ClassLabel):
            df[column] = df[column].transform(lambda i: typ.names[i])
    display(HTML(df.to_html()))

In [ ]:
# 5.e
show_random_elements(ds_train, 3)

In [ ]:
# 5.f
show_random_elements(ds_test, 1)

#### Tokenisation
###### On utilise le tokénisateur d'origine via l'objet AutoTokenizer qui le retrouve d'après le nom du modèle qui a été fourni (model checkpoint)

In [ ]:
# 5.g
# tokeniseur : on essaye le chargement de la version rapide du tokeniseur (si le temps est déjà bon, essayer la version plus lente)
tokeniseur = AutoTokenizer.from_pretrained(nom_modele, use_fast=True)

In [ ]:
# 5.h
# fonction à employer sur le dataset pour utiliser le tokeniseur
def tokenise(a_traiter):
    return tokeniseur(a_traiter["texte_base"])

###### La fonction map applique une fonction de transformation au dataset, par exemple une tokenisation
###### Les principaux arguments sont :
- la fonction de transformation
- batched : si on procède par batches de documents (1000 par 1000 par défaut, pour argument batch_size)
- num proc : mis en oeuvre si on a plusieurs processeurs, par défaut sur 1 processeur
- input_columns : liste des colonnes à traiter, par défaut toutes
- remove_columns : liste des colonnes du dataset à retirer car ne seront plus utiles par la suite

In [ ]:
# 5.i
# tokenisation sur l'unique colonne texte_base, qu'on retire par la suite
ds_texte_tokenise = ds_texte.map(tokenise, batched=True, num_proc=4, remove_columns=["texte_base"])

In [ ]:
# 5.j
# Examen du résultat
ds_texte_tokenise["train"][1]

{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1],
 'input_ids': [0, 17, 12327, 173, 375, 49387, 544, 311, 1],
 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0]}

In [ ]:
# 5.k
print("train")
len_train = len(ds_texte_tokenise["train"])
for k in range(len_train):
  print(ds_texte_tokenise["train"][k])
print("test")
len_test = len(ds_texte_tokenise["test"])
for k in range(len_test):
  print(ds_texte_tokenise["test"][k])

train
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0, 21, 6213, 57, 68, 11842, 30, 17418, 30, 1], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0, 17, 12327, 173, 375, 49387, 544, 311, 1], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0]}
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0, 19, 1841, 36915, 7177, 34, 1359, 26, 2935, 6367, 127, 19, 480, 823, 3309, 830, 1], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0, 434, 14, 35, 53, 777, 42, 15, 2717, 2005, 38, 19, 502, 1], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0, 34, 137, 19, 279, 31391, 88, 88, 1], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0]}
{'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'input_ids': [0,

#### découpage des documents en blocs de taille limitée ("phrases") en entrée du transformeur. En général la taille pour BERT est de 512 tokens.

In [ ]:
# 5.l
# voir taille maximum (nb de modèles) acceptée par 
print(tokeniseur.model_max_length)
# 
taille_bloc = 128

512


##### Fonction directement copiée / adaptée du notebook HuggingFace

In [ ]:
# 5.m
def groupe_decoupe_textes(a_traiter):
    
    # Concaténation de tous les textes du batch à traiter
    # on "somme" les listes de tokens
    # k: 'attention_mask', 'input_ids', 'token_type_ids'
    textes_concatenes = {k: sum(a_traiter[k], []) for k in a_traiter.keys()}
    
    # et on prend la longueur de l'ensemble
    long_concatenes = len(textes_concatenes[list(a_traiter.keys())[0]])
    # en fait, on découpe l'ensemble en blocs, puis pour faire simple, on ne garde pas le dernier bloc, incomplet
    # taille_bloc en contexte global
    long_concatenes = (long_concatenes // taille_bloc) * taille_bloc
    # faire le découpage par bloc
    resultat = {
        k: [t[i : i + taille_bloc] for i in range(0, long_concatenes, taille_bloc)]
        for k, t in textes_concatenes.items()
    }
    resultat["labels"] = resultat["input_ids"].copy()
    return resultat

In [ ]:
# 5.n
# transformation du dataset avec groupement / découpage en bloc des textes
ds_modele_langue = ds_texte_tokenise.map(
    groupe_decoupe_textes,
    batched=True,
    batch_size=1000,
    num_proc=4,
)

In [ ]:
# 5.o
# vérification :
tokeniseur.decode(ds_modele_langue["train"][0]["input_ids"])

In [ ]:
# 5.p
for k, x in ds_modele_langue.items(): print(k, len(x))

### Importation du modèle de langue pré-entrainé
##### Ce modèle correspond au choix du nom du modèle donné plus haut

In [ ]:
# 5.q
modele = AutoModelForMaskedLM.from_pretrained(nom_modele)

##### La fonction data_collator prends les exemplaires et les transforme en tensor, tout en faisant du masquage aléatoire. L'entraînement consiste à prédire l'élément masqué. En faisant le masquage aléatoire à ce niveau, et non en pré-processing, on s'assure de l'aléatoirité du masquage

###### 15 % des tokens sont masqués

In [ ]:
# 5.r
collateur = DataCollatorForLanguageModeling(tokenizer=tokeniseur, mlm_probability=0.15)

##### Arguments d'apprentissage

In [ ]:
# 5.s
args_apprentissage = TrainingArguments(
    "test-mlm", # et non clm ??
    evaluation_strategy = "epoch",
    learning_rate=2e-5,
    weight_decay=0.01,
)

In [ ]:
# 5.t
trainer = Trainer(
    model=modele,
    args=args_apprentissage,
    train_dataset=ds_modele_langue["train"],
    eval_dataset=ds_modele_langue["test"],
    data_collator=collateur,
)

#### Apprentissage proprement dit, cellule à grand temps d'exécution

In [ ]:
# 5.u
trainer.train()

In [ ]:
# 5.v
resultats_eval = trainer.evaluate()
print(f"Perplexité: {math.exp(resultats_eval['eval_loss']):.2f}")

### Sauvegarde du modèle pour réutilisation ultérieure
##### Dans un répertoire particulier, un répertoire par modèle (ici on suppose qu'un seul modèle suffira, pas besoin de sous-répertoires)
###### 2 fichiers :
- pytorch_model.bin : 524 Mo
- config.json

In [ ]:
# 5.w
modele.save_pretrained("/content/drive/My Drive/Datasets NLP/COVID 19")

## Création d'un modèle de topics à partir des embeddings BERT
###### En s'inspirant de ce qui a été fait dans BERTopic

#### Chaque document est découpé en blocs de taille ingérable par le transformer BERT
- le maximum est de 512 tokens tout compris
- en l'absence de découpage par phrase (pour nos sous-titres de vidéos YouTube, sauf pour une minorité d'entre elles), on va découper les documents en bloc de taille égale
- ce découpage peut avoir lieu avant ou après tokénisation BERT, on va le faire avant. Pour éliminer tous risques on va prendre des blocs de 64 tokens.
- point très important : pour le texte en entrée du tokéniseur, on ne peut pas y introduire des bigrammes/trigrammes de la forme XXXX_YYYY. Par ailleurs, BERT est plus précis si on fournit le texte complet (formes fléchies, présence des stop-words.) On utilise donc le texte d'origine. 

#### On prend chaque bloc de texte et on calcule son embedding (dans un espace de 768 dimension)
- Chaque bloc de texte est rattaché à son document d'origine.

#### On performe un regroupement (clustering) de ces blocs de texte
- les regroupements correspondent à 1 et 1 seul topic, chaque bloc de texte est dans 1 et 1 seul topic
- mais un document appartiendra dans des proportions diverses à plusieurs topics, si des blocs sont dans des topics différents

#### On crée la correspondance termes topics en utilisant les blocs de textes associés aux topic
- pour chaque topic on concatène tous les blocs de texte qui y sont rattachés, et on applique un calcul tf-idf sur chaque terme de ces textes de topic 
- il vaut mieux utiliser concaténer les textes lemmatisés et avec bigrammes/trigrammes pour une matrice termes topics utilisable 
- pour ce il faut établir la correspondance entre blocs de texte complets et blocs de texte retravaillé ; elle sera ici sommaire, mais ce n'est pas trop grave vu comme le découpage a été fait en blocs de tailles égales

#### Au final, on a les deux matrices documents X topics et topics X termes, qu'on peut utiliser comme d'autres analyses de topics sur des modèles plus classiques

###### ------ Fin de la présentation générale

#### Pour cette partie, il faut passer le notebook en mode GPU
###### Pour ce, aller dans Modifier | Paramètres du notebook, choisir GPU comme Accélérateur matériel 

### Chargement du modèle réglé finement pour étude des topics des documents

In [ ]:
# 6.a
# le tokeniseur est le tokéniseur standard
tokeniseur = AutoTokenizer.from_pretrained('flaubert/flaubert_base_uncased', use_fast=True)

In [ ]:
# 6.b
# mais le modèle est celui qui a été (ré)entraîné, réglé finement
modele = FlaubertModel.from_pretrained("/content/drive/My Drive/Datasets NLP/COVID 19", output_hidden_states = True)
# se mettre en mode exclusivement propagation avancée
# en fait, lorsqu'on vient de charger le modèle via from_pretrained, le modèle est déjà en mode "évaluation" (par opposition à "entraînement")
modele.eval()

FlaubertModel(
  (position_embeddings): Embedding(512, 768)
  (embeddings): Embedding(67542, 768, padding_idx=2)
  (layer_norm_emb): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
  (attentions): ModuleList(
    (0): MultiHeadAttention(
      (q_lin): Linear(in_features=768, out_features=768, bias=True)
      (k_lin): Linear(in_features=768, out_features=768, bias=True)
      (v_lin): Linear(in_features=768, out_features=768, bias=True)
      (out_lin): Linear(in_features=768, out_features=768, bias=True)
    )
    (1): MultiHeadAttention(
      (q_lin): Linear(in_features=768, out_features=768, bias=True)
      (k_lin): Linear(in_features=768, out_features=768, bias=True)
      (v_lin): Linear(in_features=768, out_features=768, bias=True)
      (out_lin): Linear(in_features=768, out_features=768, bias=True)
    )
    (2): MultiHeadAttention(
      (q_lin): Linear(in_features=768, out_features=768, bias=True)
      (k_lin): Linear(in_features=768, out_features=768, bias=True)
  

### Fonctions utilisées dans la boucle de constitution des embeddings des documents 

##### Création des blocs de texte
###### Comme les sous-titres sont une liste de mots sans ponctuations, on peut se contenter d'un découpage à la hâche en blocs de taille égale

In [ ]:
# 6.c
def decoupe_texte_en_blocs(texte, taille_bloc):

  l_blocs_texte = []
  # transformation du texte en liste de tokens bruts (pour le comptage des éléments du bloc, ce n'est pas la tokénization !!)
  l_tokens_bruts_texte = texte.split()
  nb_tokens_bruts = len(l_tokens_bruts_texte)
  # nb de blocs
  nb_blocs = nb_tokens_bruts // taille_bloc
  if nb_tokens_bruts % taille_bloc != 0: nb_blocs += 1
  # découpage
  for no_bloc in range(nb_blocs):
    l_tokens_bruts_bloc = l_tokens_bruts_texte[no_bloc * taille_bloc: min((no_bloc+1) * taille_bloc, nb_tokens_bruts)]
    # avec remise sous forme de chaîne de caractères
    bloc_texte = ' '.join(l_tokens_bruts_bloc)
    l_blocs_texte.append(bloc_texte)

  # on aura aussi besoin de faire une correspondance sommaire avec le texte lemmatisé et contenant les bigrammes et trigrammes
  # le découpage en bloc pourra se mesurer en nombre de blocs avec partie décimale (réelle)
  nb_blocs_dec = nb_tokens_bruts / taille_bloc

  return l_blocs_texte, nb_blocs_dec 

##### Calcul de l'embedding pour tous les blocs de texte d'un même texte général
On prend ensuite la moyenne des embeddings des 4 dernières couches
###### Si il y a trop de blocs dans un même batch cela prend toute la mémoire, on est donc contraint de découper la liste de blocs en plusieurs sous-listes

In [ ]:
# 6.d
TAILLE_GROUPE_BLOCS = 16
def calcule_embeddings(l_blocs_texte, tokeniseur, modele, nb_couches_moyennees=4):

  # on renvoie une liste de np arrays, un par bloc de texte
  l_embeddings_blocs = []
  # pour chaque bloc de texte on récupére aussi le nb de tokens avant padding
  l_nb_tokens = []

  # découper la liste des blocs de texte en 1 ou N groupes de 16 blocs maxi
  nb_blocs = len(l_blocs_texte)
  nb_groupes_blocs = nb_blocs // TAILLE_GROUPE_BLOCS
  if nb_blocs % TAILLE_GROUPE_BLOCS != 0: nb_groupes_blocs += 1

  # boucler sur la liste des groupes de bloc
  for no_groupe in range(nb_groupes_blocs):

    # obtention des blocs de texte du groupe
    l_blocs_texte_groupe = l_blocs_texte[no_groupe * TAILLE_GROUPE_BLOCS: min((no_groupe+1) * TAILLE_GROUPE_BLOCS, nb_blocs)]

    # tokénisation de la liste des (blocs de) texte
    l_blocs_tokenises_pt = tokeniseur(l_blocs_texte_groupe, padding=True, truncation=True, return_tensors="pt")

    # récupérer le nb de tokens avant padding
    for no_texte in range(len(l_blocs_texte_groupe)):
      masque_attention = l_blocs_tokenises_pt['attention_mask'][no_texte]
      nb_tokens_max = masque_attention.shape[0]
      nb_tokens = 0
      for m in range(nb_tokens_max):
        if masque_attention[m] == 1: nb_tokens += 1
        else: break
      l_nb_tokens.append(nb_tokens)

    # mettre en entrée les textes tokenizés, et récupérer la sortie
    sorties = modele(**l_blocs_tokenises_pt)

    # examen des couches cachées
    sorties_transformers = sorties['hidden_states']

    # concaténer toutes les couches
    embeddings_tokens = torch.stack(sorties_transformers, dim=0)

    # permuter les dimensions de façon à avoir l'ordre batch, tokens, couches et embeddings
    embeddings_tokens = embeddings_tokens.permute(1, 2, 0, 3)

    # on peut transformer le tout en numpy array
    mx_embeddings_tokens = embeddings_tokens.cpu().detach().numpy()

    # une stratégie classique est de moyenner les N dernières couches
    # on suppose ici que le modèle BERT est moyen et non large (13 et non 25 couches)
    ix_dernieres_couches = 13 - nb_couches_moyennees

    # boucler texte par texte (sur la dimension batch)
    nb_textes = mx_embeddings_tokens.shape[0]
    for no_texte in range(nb_textes):

      # extraire l'embedding du texte
      mx_texte_embeddings_tokens = mx_embeddings_tokens[no_texte]

      # sélection des dernières couches
      mx_sel_N_dernieres_couches = mx_texte_embeddings_tokens[:, ix_dernieres_couches:, :]

      # pour chaque token (1ère dimension) il faut sommer sur les N couches (2ème dimension), en gardant les embeddings (dernière dimension)
      # les tokens sont alors en ligne, et les embeddings toujours en colonne
      mx_somme_N_dernieres_couches = np.sum(mx_sel_N_dernieres_couches, axis=1)

      # pour un embedding de sentence, on réduit encore en sommant sur l'axe 0 (celui des tokens, qu'on élimine)
      # et on moyenne en tenant compte du nombre de tokens effectifs du texte
      mx_moyenne_sur_bloc = np.sum(mx_somme_N_dernieres_couches, axis=0) / l_nb_tokens[no_texte]

      # on peut récupérer l'embedding
      l_embeddings_blocs.append(mx_moyenne_sur_bloc)

  # on renvoie la liste des embeddings des blocs de texte, plus la liste des nb de tokens
  return l_embeddings_blocs, l_nb_tokens


In [ ]:
# 6.e
# Eléments de dimensionnement
NB_DIM_EMBEDDINGS = 768
TAILLE_BLOC = 96
NB_COUCHES = 4

##### Créer les conteneurs suivants :
- une matrice des embeddings (blocs de texte X dimensions d'embedding)
- un dictionnaire faisant la correspondance entre blocs de textes et textes (avec position) et le dictionnaire inverse listant blocs du texte dans leur ordre (plus le nb réel/décimal de blocs pour découpage rapide)
- une liste des nb de tokens des blocs de texte, et une approximation de leur contribution au textes dont ils sont issus 

###### Alors qu'il n'y a que 1342 sous-titres de vidéos, le tout prend 2 heures !!! - réduire à 300 vidéos...

In [ ]:
# df_docs = df_videos
df_docs = df_videos.sample(300)
df_docs.reset_index(inplace=True)

In [ ]:
# 6.f
mx_blocs_embeddings = np.empty((0,NB_DIM_EMBEDDINGS))
d_blocs_docs = {}
d_docs_blocs = {}
l_nb_tokens_blocs = []

In [ ]:
# 6.g
# Création des conteneurs des blocs de texte

no_bloc_corpus = 0

for no_doc, doc in df_docs['texte_base'].iteritems():

  # découpage du doc en blocs
  l_blocs_doc, nb_blocs_dec = decoupe_texte_en_blocs(doc, TAILLE_BLOC)
  nb_blocs = len(l_blocs_doc)
  
  # initialisation dic au niveau doc
  d_docs_blocs[no_doc] = ([], nb_blocs_dec)

  # obtention des embeddings
  l_embeddings_blocs, l_nb_tokens = calcule_embeddings(l_blocs_doc, tokeniseur, modele, nb_couches_moyennees=NB_COUCHES)

  # compléter éléments
  for no_bloc_doc, embedding_bloc in enumerate(l_embeddings_blocs):

    # bloc -> doc
    d_blocs_docs[no_bloc_corpus] = (no_doc, no_bloc_doc)
    # doc -> bloc
    d_docs_blocs[no_doc][0].append(no_bloc_corpus)

    # ajouter une ligne par bloc à la matrice bloc X embeddings
    mx_blocs_embeddings = np.append(mx_blocs_embeddings, [embedding_bloc], axis=0)

    # et nombre de tokens (au cas où)
    l_nb_tokens_blocs.append(l_nb_tokens[no_bloc_doc])

    no_bloc_corpus += 1

  # avancement
  if (no_doc+1) % 20 == 0: print("%d embeddings obtenus pour %d docs" % (no_bloc_corpus, no_doc+1))

# nombre total de blocs sur tout le corpus
nb_bloc_corpus = no_bloc_corpus


240 embeddings obtenus pour 20 docs
533 embeddings obtenus pour 40 docs
960 embeddings obtenus pour 60 docs
1123 embeddings obtenus pour 80 docs
1771 embeddings obtenus pour 100 docs
2341 embeddings obtenus pour 120 docs
2664 embeddings obtenus pour 140 docs
3040 embeddings obtenus pour 160 docs
3233 embeddings obtenus pour 180 docs
3552 embeddings obtenus pour 200 docs
3830 embeddings obtenus pour 220 docs
4304 embeddings obtenus pour 240 docs
4460 embeddings obtenus pour 260 docs
4648 embeddings obtenus pour 280 docs
4945 embeddings obtenus pour 300 docs


##### Réduction de dimensionnalité des embeddings avant le clustering
- le nombre de dimensions cible recommandé (BERTopic) est de 5, à ajuster selon le nombre de topics trouvé ultérieurement et la proportion de blocs non regroupés.
- si le nombre de dimensions est trop élevé, il y a difficultés pour regrouper des documents (blocs) qui pourraient en fait l'être
- si le nombre est trop faible, les topics ne sont pas assez discriminant

In [ ]:
# 6.h
# Nous utilisons ici les paramètres choisis par BERTopic
NB_VOISINS = 15
#NB_DIMS_REDUITES = 5 # Il faut augmenter cette dimension pour obtenir un nombre raisonnable de topics
NB_DIMS_REDUITES = 10
DISTANCE_MIN = 0.0 # plutôt que 0.1
METRIQUE_UMAP = 'cosine' # plutôt que 'euclidean'

In [ ]:
# 6.i
reducteur_umap = umap.UMAP(n_neighbors=NB_VOISINS, min_dist=DISTANCE_MIN, n_components=NB_DIMS_REDUITES, metric=METRIQUE_UMAP)
mx_blocs_reduite = reducteur_umap.fit_transform(mx_blocs_embeddings)

##### Effectuer le regroupement par HDBScan
- min_cluster_size : quelle est la taille minimum (nb de blocs) que doit avoir un cluster pour pouvoir exister ? Ce qui joue sur les décisions de séparation de clusters. Plus la taille est grande, moins il y aura de petits topics.
- metric : 'euclidean' est la métrique par défaut. Pas de cosinus, car ne fait pas directement sens pour un algorithme de regroupement.

In [ ]:
# 6.j
# Nous utilisons ici les paramètres choisis par BERTopic
TAILLE_MIN_TOPIC = 10
METRIQUE_HDBSCAN = 'euclidean'

In [ ]:
# 6.k
regroupeur_HDBSCAN = hdbscan.HDBSCAN(min_cluster_size=TAILLE_MIN_TOPIC, metric=METRIQUE_HDBSCAN)
regroupeur_HDBSCAN.fit(mx_blocs_reduite)

HDBSCAN(algorithm='best', allow_single_cluster=False, alpha=1.0,
        approx_min_span_tree=True, cluster_selection_epsilon=0.0,
        cluster_selection_method='eom', core_dist_n_jobs=4,
        gen_min_span_tree=False, leaf_size=40,
        match_reference_implementation=False, memory=Memory(location=None),
        metric='euclidean', min_cluster_size=10, min_samples=None, p=None,
        prediction_data=False)

###### Observation du nombre de clusters, et pour chacun de ses clusters, déterminer fiabilité moyenne d'apparetance (on pourrait également examiner distribution plus en profondeur)

In [ ]:
# 6.l
# calcul du nb de topics
# obtenir le nombre de groupes trouvés (de 0 à N)
nb_topics = regroupeur_HDBSCAN.labels_.max() + 1
# puis ajouter le pseudo-topic -1 (pas de regroupements)
nb_topics += 1
print(nb_topics)

56


###### Par la suite, on va assigner au pseudo-topic le numéro et le label 0, et aux autres de 1 à N(+1)

In [ ]:
# 6.m
# et créer la liste des nombres et probabilités de cluster
d_topics = {}
for no_topic, proba_topic in zip(regroupeur_HDBSCAN.labels_, regroupeur_HDBSCAN.probabilities_):
  if (no_topic+1) not in d_topics:
    d_topics[no_topic+1] = (1, proba_topic)
  else:
    d_topics[no_topic+1] = (d_topics[no_topic+1][0] + 1, d_topics[no_topic+1][1] + proba_topic)
l_topics = []
for no_topic, (nb_ds_topic, somme_proba_topics) in d_topics.items():
  l_topics.append((no_topic, nb_ds_topic, somme_proba_topics / nb_ds_topic))
# trier selon les numéros de topic
l_topics.sort(key=lambda x: x[0])

In [ ]:
# 6.n
# examen rapide de la qualité des topics
for topic in l_topics: print(topic)

(0, 2441, 0.0)
(1, 20, 0.802483813494678)
(2, 30, 0.6454298501062358)
(3, 35, 0.6664792732456201)
(4, 22, 0.6647984865142086)
(5, 19, 0.7833656356532361)
(6, 17, 0.8964217552682427)
(7, 13, 0.9917366236666246)
(8, 171, 0.5308227346807721)
(9, 27, 0.7794319402976798)
(10, 13, 0.9974776084717123)
(11, 18, 0.9879106872021506)
(12, 60, 0.6669152977076268)
(13, 78, 0.4608623016653843)
(14, 594, 0.912220027028298)
(15, 20, 0.8659250898561552)
(16, 10, 1.0)
(17, 10, 1.0)
(18, 11, 0.9998489602727753)
(19, 22, 0.9062421448251108)
(20, 18, 0.9408692247938366)
(21, 13, 0.9946469041128272)
(22, 11, 0.9996441862585947)
(23, 12, 0.9922704358346186)
(24, 12, 0.9959916489968107)
(25, 27, 0.9292913742288037)
(26, 10, 1.0)
(27, 37, 0.9030926318251342)
(28, 27, 0.9412274840974315)
(29, 11, 0.9982169235329846)
(30, 22, 0.8902069157749953)
(31, 10, 1.0)
(32, 45, 0.5017495210815157)
(33, 64, 0.7110300243420653)
(34, 15, 0.9875890275690898)
(35, 11, 0.9988075745424979)
(36, 13, 0.9954503936048403)
(37, 61, 0

##### Création des matrices docs X topics et topics X termes

###### Pour la matrice docs X topics, il faut pour chaque doc retrouver les topics de ses blocs de texte, et y associer les différents topics en tenant compte de la longueur des blocs

In [ ]:
# 6.o
mx_docs_topics = np.zeros((nb_docs, nb_topics))

for no_doc, (l_blocs_doc, nb_blocs_dec) in d_docs_blocs.items():

  mx_topics = np.zeros((nb_topics))
  
  # parcourir les blocs du doc
  for no_bloc_doc, no_bloc_corpus in enumerate(l_blocs_doc):
    # décalage de 1 pour sans topic à 0 et les autres à partir de 1
    no_topic_bloc = regroupeur_HDBSCAN.labels_[no_bloc_corpus] + 1
    proba_topic_bloc = regroupeur_HDBSCAN.probabilities_[no_bloc_corpus]
    # poids du bloc, 1, sauf si dernier bloc avec résidus du découpage
    poids_bloc = 1 if (no_bloc_doc + 1 < nb_blocs_dec) else (nb_blocs_dec - no_bloc_doc)
    # incrémenter le poids du topic correspondant
    # en fait, il peut arriver que tous les blocs d'un même doc aient une probabilité nulle,
    # ne pas tenir compte de cette dernière, en tous cas si la somme est nulle, et
    # en attendant, dans tous les cas de figure
    #mx_topics[no_topic_bloc] += poids_bloc * proba_topic_bloc
    mx_topics[no_topic_bloc] += poids_bloc

  # normaliser tous les poids de façon à ce que la somme fasse 1
  #somme_topics = np.sum(mx_topics)
  somme_topics = nb_blocs_dec
  #print(no_doc, somme_topics)
  mx_topics = mx_topics / somme_topics
  
  # porter dans la matrice docs X topics
  mx_docs_topics[no_doc] = mx_topics

0 0.34375
1 2.0416666666666665
2 2.21875
3 2.9791666666666665
4 3.7916666666666665
5 8.1875
6 3.3854166666666665
7 5.927083333333333
8 1.4270833333333333
9 12.197916666666666
10 15.1875
11 6.927083333333333
12 29.385416666666668
13 1.9479166666666667
14 15.208333333333334
15 29.8125
16 61.947916666666664
17 7.114583333333333
18 1.75
19 18.135416666666668
20 2.6354166666666665
21 18.0
22 2.4166666666666665
23 41.0
24 18.833333333333332
25 3.4479166666666665
26 3.8541666666666665
27 24.510416666666668
28 2.375
29 3.21875
30 3.1875
31 1.90625
32 3.6145833333333335
33 27.104166666666668
34 2.03125
35 5.291666666666667
36 26.416666666666668
37 45.020833333333336
38 3.28125
39 44.958333333333336
40 4.875
41 45.291666666666664
42 2.90625
43 35.3125
44 3.1041666666666665
45 78.02083333333333
46 5.34375
47 8.010416666666666
48 48.75
49 1.8333333333333333
50 11.15625
51 3.03125
52 0.875
53 18.739583333333332
54 41.4375
55 2.3854166666666665
56 16.010416666666668
57 72.72916666666667
58 4.5625
59

##### Pour la matrice topics X termes on procède comme BERTopic
- on aggrège les textes des différents blocs appartenant à un même topic
- on applique dessus une métrique c-Tf-Idf 

###### Aggrégation des textes des différents blocs appartenant à un même topic
 - on utilise la colonne des textes réduits, lemmatisés et avec bigrammmes / trigrammes en faisant la supposition que le découpage en bloc est (quasiment) proportionnel entre texte de départ et texte lemmatisé


In [ ]:
# 6.p
# on crée les listes de tokens associées à chaque topic, on les initialise à vide
l_topics_l_tokens = []
for no_topic in range(nb_topics):
  l_topics_l_tokens.append([])

# boucler sur les documents
for no_doc, texte in df_docs['textes_sous_titres_avec_bigrammes'].iteritems():

  # découper le texte selon les blocs
  nb_blocs = len(d_docs_blocs[no_doc][0])
  nb_blocs_dec = d_docs_blocs[no_doc][1]
  l_tokens = texte.split()
  nb_tokens = len(l_tokens)

  # retrouver les blocs
  taille_bloc = int(nb_tokens / nb_blocs_dec)
  for no_bloc in range(nb_blocs):
    # tokens du bloc
    l_tokens_bloc = l_tokens[no_bloc * taille_bloc: min((no_bloc+1)* taille_bloc, nb_tokens)]
    # topic du bloc
    no_bloc_corpus = d_docs_blocs[no_doc][0][no_bloc]
    no_topic = regroupeur_HDBSCAN.labels_[no_bloc_corpus] + 1 
    # concaténation sur la liste générale des tokens du topic correspondant
    l_topics_l_tokens[no_topic].extend(l_tokens_bloc) 


In [ ]:
# 6.q
# comme on utilise le vectorisateur de scikit-learn, on va devoir transformer les listes de tokens en chaînes de caractère
l_topics_l_textes = []
for no_topic in range(nb_topics):
  l_topics_l_textes.append(' '.join(l_topics_l_tokens[no_topic]))

###### Métrique c-tf-idf : analogue au calcul de tf-idf mais sur autant de documents que de topics.
Cependant: 
1) dans tf on normalise la fréquence en la divisant par le nombre de tokens dans le document (le topic)
2) dans idf, on considère non le nombre de documents / topics, mais le nombre de documents d'origine (de blocs), et qu'on divise non par le nombre de document dans lesquels les termes apparaissent mais par la fréquence totale du terme dans l'ensemble des documents

In [ ]:
# 6.r
# sélection des termes dans les bonnes fréquences et dans les bon nombres
MAX_DFREQ = 0.95
MIN_DFREQ = 5
NB_TOTAL_TERMES = 10000

# et comme le travail de split a été effectué, surcharger le tokénisateur par défaut avec un split simple
def simple_split(s):
  return s.split()

In [ ]:
 # 6.s
 # travailler à partir d'une matrice des comptages
 # on va éliminer comme à l'habitude les termes les plus fréquents et les plus rares, 
 # et garder un nombre de termes raisonnable 
 freqs = CountVectorizer(max_df=MAX_DFREQ, min_df=MIN_DFREQ, max_features=NB_TOTAL_TERMES, 
                         tokenizer=simple_split)

 # vectorisation, matrice topics X termes          
 mx_f_topics_termes = freqs.fit_transform(l_topics_l_textes)
 ix_termes = freqs.get_feature_names()
 nb_termes = len(ix_termes)

In [ ]:
# 6.t
# adapter la matrice pour en faire une matrice c-tf-idf

# il y a besoin pour chaque topic de la somme des fréquences de tous les termes
mx_freq_topics = np.zeros((nb_topics))
for no_topic in range(nb_topics):
  mx_freq_topics[no_topic] = len(l_topics_l_tokens[no_topic])

# et pour chaque terme de la somme de ses fréquences sur tous les topics
mx_freq_termes = np.zeros((nb_termes))
for no_terme in range(nb_termes):
  mx_freq_termes[no_terme] = mx_f_topics_termes[:,no_terme].sum()

# d'où la matrice définitive
mx_topics_termes = np.zeros((nb_topics, nb_termes))
for no_topic in range(nb_topics):
  for no_terme in range(nb_termes):
      tf = mx_f_topics_termes[no_topic, no_terme] / mx_freq_topics[no_topic]
      idf = math.log(nb_bloc_corpus / mx_freq_termes[no_terme],2)
      mx_topics_termes[no_topic, no_terme] = tf * idf

#### On a maintenant les éléments nécessaires pour faire des analyses de topics
###### On va se contenter de lister les topics et leurs termes les plus importants

In [ ]:
# 7.a
# trouver la liste des top termes (plus leurs poids) d'un topic
def topic_top_termes(mx_topics_termes, ix_termes, no_topic, nb_top_termes):
  v_topic_termes = mx_topics_termes[no_topic]    
  return [(ix_termes[no_terme], v_topic_termes[no_terme])\
          for no_terme in np.argsort(v_topic_termes)[:-nb_top_termes-1:-1]]

In [ ]:
# 7.b
# trouver la liste des top docs (plus leurs poids) d'un topic
def topic_top_docs(mx_docs_topics, no_topic, nb_top_docs):
  v_topic_docs = mx_docs_topics[:,no_topic]    
  return [(no_doc, v_topic_docs[no_doc])\
          for no_doc in np.argsort(v_topic_docs)[:-nb_top_docs-1:-1]]

In [ ]:
# 7.c
# trouver la masse totale sur les docs d'un topic
def topic_masse_totale(mx_docs_topics, no_topic):
  return mx_docs_topics[:,no_topic].sum() 

In [ ]:
# 7.d
# dériver certaines informations sur le modèle de topics
def informations_derivees(mx_docs_topics, mx_topics_termes):

  # liste des topics avec un nom parlant dérivé des noms des termes les plus représentatifs,
  # plus le poids des topics 
  nb_topics = mx_topics_termes.shape[0]
  l_topics = []
  for no_topic in range(nb_topics):
    topic_attributs = {}
    top_2_termes = topic_top_termes(mx_topics_termes, ix_termes, no_topic, 2)
    topic_attributs['nom'] = "%s %s" % (top_2_termes[0][0], top_2_termes[1][0])
    topic_attributs['masse'] = topic_masse_totale(mx_docs_topics, no_topic)
    l_topics.append(topic_attributs)

  return l_topics

In [ ]:
# 7.e
l_topics = informations_derivees(mx_docs_topics, mx_topics_termes)

In [ ]:
# 7.f
# liste textuelle des topics
def montre_liste_topics(l_topics, mx_topics_termes, ix_termes,  
                        nb_top_termes=10, avec_masses_topics=False, avec_masses_termes=False):
        
  for no_topic in range(len(l_topics)):

    # topic en tant que tel
    s_nom_topic = "%d - %s" % (no_topic+1, l_topics[no_topic]['nom'])
    s_masse_topic = ""
    if avec_masses_topics == True:
      s_masse_topic = " (masse %.3f)" % (l_topics[no_topic]['masse'])
    s_topic = "%s%s : " % (s_nom_topic, s_masse_topic)
        
    # et les N top termes
    s_top_termes = ""
    top_termes = topic_top_termes(mx_topics_termes, ix_termes, no_topic, nb_top_termes)
    for ix, (terme, masse) in enumerate(top_termes):
      s_masse_terme = ""
      if avec_masses_termes == True:
        s_masse_terme = ": %.3f" % masse
      s_sep_terme = ", "
      if ix == nb_top_termes - 1: s_sep_terme = ""
      s_top_termes += "%s%s%s" % (terme, s_masse_terme, s_sep_terme) 
        
    # sortie finale
    print("%s%s" % (s_topic, s_top_termes))

In [ ]:
# 7.g
montre_liste_topics(l_topics, mx_topics_termes, ix_termes, 
                    nb_top_termes=10, avec_masses_topics=True, avec_masses_termes=False)

1 - yer faire (masse 132.173) : yer, faire, vouloir, dire, voir, savoir, pouvoir, aller, chose, prendre
2 - candidat tour (masse 2.258) : candidat, tour, écologiste, socialiste, alliance, tête, verts, abstention, liste, voix
3 - martine tribunal (masse 1.406) : martine, tribunal, mourir, homme, ferme, fou, of, invisible, auteur, offre
4 - immobilier crédit (masse 4.772) : immobilier, crédit, fin, revenu, britannique, occasion, pourcent, marché, déficit, crise
5 - joueur club (masse 5.218) : joueur, club, match, saison, lens, championnat, om, jouer, reporter, stade
6 - application téléphone (masse 2.799) : application, téléphone, installer, apple, antoine, télécharger, montre, analyser, automatiquement, nouveau
7 - indépendance solidaire (masse 1.185) : indépendance, solidaire, bâtir, nation, européen, écologique, nom, priorité, souffrir, mieux
8 - compte banque (masse 0.714) : compte, banque, dédier, professionnel, activité, indépendant, bancaire, législatif, obligatoire, recours
9 - v